processing/src/main/java/io/druid/query/filter/LikeDimFilter.java

```diff
@@ -0,0 +1,362 @@
+import com.google.common.io.BaseEncoding;
+import com.google.common.primitives.Chars;
+import io.druid.java.util.common.StringUtils;
+import io.druid.query.extraction.ExtractionFn;
+import io.druid.segment.filter.LikeFilter;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+import java.util.regex.Pattern;
+
+public class LikeDimFilter implements DimFilter
+{
+  // Regex matching characters that are definitely okay to include unescaped in a regex.
+  // Leads to excessively paranoid escaping, although shouldn't affect runtime beyond compiling the regex.
+  private static final Pattern DEFINITELY_FINE = Pattern.compile("[\\w\\d\\s-]");
+  private static final String WILDCARD = ".*";
+
+  private final String dimension;
+  private final String pattern;
+  private final Character escapeChar;
+  private final ExtractionFn extractionFn;
+  private final LikeMatcher likeMatcher;
+
+  @JsonCreator
+  public LikeDimFilter(
+      @JsonProperty("dimension") final String dimension,
+      @JsonProperty("pattern") final String pattern,
+      @JsonProperty("escape") final String escape,
+      @JsonProperty("extractionFn") final ExtractionFn extractionFn
+  )
+  {
+    this.dimension = Preconditions.checkNotNull(dimension, "dimension");
+    this.pattern = Preconditions.checkNotNull(pattern, "pattern");
+    this.extractionFn = extractionFn;
+
+    if (escape != null && escape.length() != 1) {
+      throw new IllegalArgumentException("Escape must be null or a single character");
+    } else {
+      this.escapeChar = (escape == null || escape.isEmpty()) ? null : escape.charAt(0);
+    }
+
+    this.likeMatcher = LikeMatcher.from(pattern, this.escapeChar, extractionFn == null);
+  }
+
+  public static class LikeMatcher
+  {
+    public enum SuffixStyle
+    {
+      MATCH_ANY,
+      MATCH_EMPTY,
+      MATCH_PATTERN
+    }
+
+    // Prefix that matching strings will always start with. Used for helping with index lookups.
+    private final String prefix;
+
+    // Strings match if:
+    //  (a) suffixStyle is MATCH_ANY and they start with "prefix"
+    //  (b) suffixStyle is MATCH_EMPTY and they start with "prefix" and contain nothing after prefix
+    //  (c) suffixStyle is MATCH_PATTERN and the string matches "pattern"
+    private final SuffixStyle suffixStyle;
+    private final Pattern pattern;
+
+    private LikeMatcher(
+        final String prefix,
+        final SuffixStyle suffixStyle,
+        final Pattern pattern
+    )
+    {
+      this.prefix = Strings.nullToEmpty(prefix);
+      this.suffixStyle = Preconditions.checkNotNull(suffixStyle, "suffixStyle");
+      this.pattern = Preconditions.checkNotNull(pattern, "pattern");
+    }
+
+    public static LikeMatcher from(
+        final String likePattern,
+        @Nullable final Character escapeChar,
+        final boolean extractPrefix
+    )
+    {
+      final StringBuilder prefix = new StringBuilder();
+      final StringBuilder suffixRegex = new StringBuilder();
```

Comment:
I think now this method could be simplified. Instead of accumulating `suffixRegex`, it could just assign suffixStyle variable, and make a simple constructor call in the end instead of if-else chain.